usaspending_api/search/v2/views/spending_by_category_views/base_spending_by_category.py

```diff
@@ -0,0 +1,153 @@
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from typing import List
+
+from django.conf import settings
+from django.db.models import QuerySet, Sum
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from usaspending_api.awards.v2.filters.sub_award import subaward_filter
+from usaspending_api.awards.v2.filters.view_selector import spending_by_category_view_queryset
+from usaspending_api.common.api_versioning import api_transformations, API_TRANSFORM_FUNCTIONS
+from usaspending_api.common.cache_decorator import cache_response
+from usaspending_api.common.elasticsearch.search_wrappers import TransactionSearch
+from usaspending_api.common.experimental_api_flags import is_experimental_elasticsearch_api
+from usaspending_api.common.helpers.generic_helper import get_simple_pagination_metadata, get_time_period_message
+from usaspending_api.common.query_with_filters import QueryWithFilters
+from usaspending_api.common.validator.award_filter import AWARD_FILTER
+from usaspending_api.common.validator.pagination import PAGINATION
+from usaspending_api.common.validator.tinyshield import TinyShield
+from usaspending_api.search.v2.elasticsearch_helper import get_number_of_unique_terms
+
+logger = logging.getLogger("console")
+
+
+@dataclass
+class Category:
+    name: str
+    primary_field: str
+
+
+@api_transformations(api_version=settings.API_VERSION, function_list=API_TRANSFORM_FUNCTIONS)
+class BaseSpendingByCategoryViewSet(APIView, metaclass=ABCMeta):
+    """
+    Base class inherited by the different category endpoints.
+    """
+
+    category: Category
+    elasticsearch: bool
+    filters: dict
+    obligation_column: int
+    pagination: dict
+    subawards: bool
+
+    @cache_response()
+    def post(self, request: Request) -> Response:
+        models = [
+            {"name": "subawards", "key": "subawards", "type": "boolean", "default": False, "optional": True},
+        ]
+        models.extend(copy.deepcopy(AWARD_FILTER))
+        models.extend(copy.deepcopy(PAGINATION))
+
+        validated_payload = TinyShield(models).block(request.data)
+
+        validated_payload["elasticsearch"] = is_experimental_elasticsearch_api(request)
+
+        return Response(self.handle_business_logic(validated_payload))
+
+    def handle_business_logic(self, validated_payload: dict) -> dict:
+
+        self.filters = validated_payload.get("filters", {})
+        self.elasticsearch = validated_payload.get("elasticsearch")
+        self.subawards = validated_payload["subawards"]
+        self.pagination = {
+            "page": validated_payload["page"],
+            "limit": validated_payload["limit"],
+            "lower_limit": (validated_payload["page"] - 1) * validated_payload["limit"],
+            "upper_limit": validated_payload["page"] * validated_payload["limit"] + 1,
+        }
+
+        if self.elasticsearch and not self.subawards:
```

Comment:
Could this conditional be simplified?

if self.subawards:
....
elif self.elasticsearch:
....
else:
....

Reply:
I'm not sure if subawards should be in this design at all, since that's very use case specific. If this chunk of logic is going to be different for each kind of category, this is probably a place to have another abstract method for the implementers.

Reply:
You don't expect the ability to surface sub-award data for each/most categories?

The abstract methods are called here depending on the request JSON and headers (subawards vs prime awards, elasticsearch vs postgres), I'm not tracking the benefit for making this re-defined in each child class.

Reply:
Right, I forget that subawards is a whole tab sometimes. Given the logical seperation here, I would argue in favor of keeping two levels and simplifying the first conditional:
```
if self.elasticsearch:
...
else:
    if self.subawards:
    ...
    else:
```
to isolate the es logic more easily.

Reply:
Funny, I thought my suggestion was following the logical separation better. Essentially, sub-awards will always be separate from prime-awards no matter where the data are stored. The Postgres/Elasticsearch separated business logic is temporary and is still searching and surfacing the same data.

If possible, it is always a good idea to "flatten" the code. It wasn't obvious what the nested conditionals added